language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -330,46 +350,55 @@ withScopedArgIn funVarEnv argVarEnv arg            k = do
-    withScopedArgIn funVarEnv argVarEnv arg $ \arg' ->
-        let term = Apply (memoryUsage () <> memoryUsage fun <> memoryUsage arg') fun arg'
-        in case termAsPrimIterApp term of
-            Nothing                       ->
-                throwingWithCause _MachineError NonPrimitiveApplicationMachineError $ Just (void term)
-            Just (IterApp headName spine) -> do
-                constAppResult <- applyStagedBuiltinName headName spine
-                case constAppResult of
-                    ConstAppSuccess res -> computeCek con res
-                    ConstAppStuck       -> returnCek con term
-
--- | Apply a 'StagedBuiltinName' to a list of 'Value's.
-applyStagedBuiltinName
-    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => StagedBuiltinName
-    -> [WithMemory Value uni]
-    -> CekM uni (ConstAppResult uni ExMemory)
-applyStagedBuiltinName n@(DynamicStagedBuiltinName name) args = do
-    DynamicBuiltinNameMeaning sch x exX <- lookupDynamicBuiltinName name
-    applyTypeSchemed n sch x exX args
-applyStagedBuiltinName (StaticStagedBuiltinName name) args = do
-    params <- builtinCostParams
-    applyBuiltinName params name args
+
+-- If it's a standard builtin, all arguments should be (built-in)
+-- values and this function will do nothing.  Can we use traverse or
+-- something for this? We're simultaneously mapping across the args
+-- and folding along the environment
+getScopedArgs  -- We could perhaps fold this into `compute`: more efficient, but more complex?
+               -- Let's leave it like this until we're sure it's correct.
+    :: [Closure uni]           -- original arguments
+    -> [WithMemory Value uni]  -- adjusted arguments
+    -> VarEnv uni              -- enviroment containing bindings for new variables
+    -> CekM uni ([WithMemory Value uni], VarEnv uni)
+getScopedArgs [] newArgs env = pure (reverse newArgs, env)
+getScopedArgs (Closure argVarEnv arg : args) newArgs env =
+    case arg of
+      Constant{} -> getScopedArgs args (arg : newArgs) env
+      _ -> do
+        argName <- freshName "arg"
+        let cost = memoryUsage ()
+            newArg = Var cost argName
+            newEnv = extendVarEnv argName arg argVarEnv env
+        getScopedArgs args (newArg : newArgs) newEnv
+
+applyBuiltin :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
+                 => VarEnv uni
+                 -> ann
+                 -> Context uni
+                 -> BuiltinName
+                 -> [Type TyName uni ExMemory]
+                 -> [Closure uni]
+                 -> CekM uni (Term TyName Name uni ())
+applyBuiltin funEnv ann ctx bn tys args =
+    -- If there are too many/few types this should be caught by the typechecker.
+    let term = ApplyBuiltin () bn (fmap void tys) (fmap (void . _closureValue) args) -- For error messages
+    in do
+      -- Do we have to be more careful for dynamic names?
```

Comment:
Do you have something in mind?

Reply:
I'm not sure: I think that was just to remind me to go back and think more carefully.